File: 0242-valid-anagram/0242-valid-anagram.rs
use std::collections::HashMap;

impl Solution {
    pub fn is_anagram(s: String, t: String) -> bool {
        if s.len() != t.len() {
            return false;
        }

        let mut letters: HashMap<char, i32> = HashMap::new();
        for (a, b) in s.chars().zip(t.chars()) {
            *letters.entry(a).or_default() += 1;
            *letters.entry(b).or_default() -= 1;
        }

        letters.into_values().all(|count| count == 0)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_1() {
        assert!(Solution::is_anagram(
            "anagram".to_string(),
            "nagaram".to_string()
        ));
    }

    #[test]
    fn test_2() {
        assert!(!Solution::is_anagram("rat".to_string(), "car".to_string()));
    }
}

File: 0242-valid-anagram/0242-valid-anagram.py
class Solution:
    def isAnagram(self, s: str, t: str) -> bool:
        if len(s) != len(t):
            return False

        letters = {}
        for value in s:
            letters[value] = letters.get(value, 0) + 1
            
        for value in t:
            letters[value] = letters.get(value, 0) - 1
        
        for value in letters.values():
            if value != 0:
                return False

        return True
        